### PR TITLE
Add user store to packages/data-stores and use wpcom-proxy-request for API calls

### LINKED ISSUE #38794: [Gutenboarding]: Frontend account creation request/response handling
Subtask of EPIC #38701

## Create an account

This task involves adding code to the Gutenboarding/data-store architecture to call the `users/new` endpoint and handle the response.

## Technical details

From an API perspective, we could reuse our work from passwordless signup (paWIov-6C-p2).

At the moment, we generate a username based on arrays of random words, e.g., `rabbitsaturnpizza`. If a user has entered information in the site title field however, this should take priority as a base for a user name. If a user has entered information in the site vertical field, this should take second priority. If no further information is available, then we can fall back on the existing random generator. 

We can probably use the existing `extra` field. See: https://github.com/Automattic/wp-calypso/blob/master/client/signup/steps/user/index.jsx#L199

For the backend task see #38795

## Tasks

- [ ] Investigate what we need to do (if anything) to hit the WordPress.com API without `lib/wp`. Related comments: https://github.com/Automattic/wp-calypso/issues/38703#issuecomment-572479664
- [ ] It seems like we could integrate HTTP `POST` request and handle the response in [`@automattic/data-stores` package](https://github.com/Automattic/wp-calypso/blob/6f8da7a/packages/data-stores).

### PR DESCRIPTION
**Pull request by andrewserong**
The object of this PR is to add a `user` data store that handles checking for a logged in user, and account creation with fetching and error states, using patterns from the `@wordpress/data` package. A number of selectors and actions are provided to hopefully allow for creating simple sign up interfaces. I've included a readme with an example of how to use the data store.

#### Changes proposed in this Pull Request

* Add `wpcom-proxy-request` to `packages/data-stores` for making proxied API calls to wp.com
* Add a simple `currentUser` object
* Add a simple `newUser` object
* Add a Promise-based wrapper for `wpcom-proxy-request` calls [similar to what's used in `wpcom.js`](https://github.com/Automattic/wpcom.js/blob/94edca17553d597f9de68fc03efe488158a5babd/lib/util/send-request.js#L88)
* Add `debug`
* Add actions and selectors for checking if a user is logged in, and for creating a new user using passwordless login, with error and fetching state

This PR supersedes the other similar experiments in #38679 and #38739

#### Still to be done

* [x] ~Add fetching state~
* [x] ~Clean up types a bit more (error object is currently set to `object`)~
* [x] ~Remove magic strings (put the client id and 'secret' somewhere else)~

#### Testing instructions

* This PR adds an extra data store but doesn't yet use it anywhere so there are no regressions to test against.
* To test the behaviour of the store, follow the included `README.md` file which has an example signup form you can use.
* To see debug logs of the `wpcom-proxy-request` responses, run the following in your browser console `localStorage.setItem( 'debug', 'data-stores:*' );`

Fixes #38794

### CODE REVIEW

File: packages/data-stores/src/user/index.ts
Comment (andrewserong) [reply packages/data-stores/src/user/index.ts]: Good idea, thanks for the extra background @ramonjd! I suppose it'd make sense to use a similar pattern as in https://github.com/automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/lib/wp/browser.js#L28 — I'll have a play.

File: client/landing/gutenboarding/components/header/index.tsx
Comment (p-jackson) [inline client/landing/gutenboarding/components/header/index.tsx]: Nit: maybe move this so it's right beside the `DOMAIN_SUGGESTIONS_STORE` declaration?

File: packages/data-stores/src/user/actions.ts
Comment (p-jackson) [inline packages/data-stores/src/user/actions.ts]: I'm not a fan of having to specify `as const` in all these action creators but I think this pattern was copied from other data stores. I might have to come up with a proposal for how we can rework the types if I want it to be something different :)

File: packages/data-stores/src/user/controls.ts
Comment (p-jackson) [inline packages/data-stores/src/user/controls.ts]: These usually come from `development.json` et al. Since Gutenberg is it's own app (I think @sirreal even wants to move it to `/app`) I guess it needs it's own app config? Doesn't seem like it's something it should share with Calypso.

Comment (p-jackson) [reply packages/data-stores/src/user/controls.ts]: We don't even need anything as sophisticated as Calypso's config-per-environment. `client_id` and `client_secret` are the same in all environments, so just hard-coding them in a constants file should be fine.

To keep the config out of the package I guess they could just be included in `CreateAccountParams`? KISS?
It'd be cool if `User.register()` accepted parameters so the config could be passed in when initing the store.

File: packages/data-stores/src/user/constants.ts
Comment (p-jackson) [inline packages/data-stores/src/user/constants.ts]: Trying to decide if this store should be called `user` or `auth` 🤔

File: packages/data-stores/src/user/index.ts
Comment (p-jackson) [inline packages/data-stores/src/user/index.ts]: Silly me, `User.register()` can take parameters 😆 
I reckon pass the `client_id` and `client_secret` in here. Then you can use a control factory to create a controls that uses the correct config.

File: packages/data-stores/src/user/index.ts
Comment (p-jackson) [inline packages/data-stores/src/user/index.ts]: Nice!

File: packages/data-stores/src/user/reducer.ts
Comment (p-jackson) [inline packages/data-stores/src/user/reducer.ts]: Doesn't feel like this is going to scale as a reducer takes more actions, but I think this is a pattern that's copied from other data-stores?

File: packages/data-stores/src/utils/wpcom-wrapper.ts
Comment (ramonjd) [inline packages/data-stores/src/utils/wpcom-wrapper.ts]: I'm more asking for advice here as I'm not sure: would it make sense to add type definitions to `wpcomProxyRequest` in the global declarations? 

That way the parser will complain if `wpcomProxyRequest` doesn't receive what it expects.

For example, in `packages/data-stores/global.d.ts`:

```
declare module 'wpcom-proxy-request' {
   // import only works if we add "baseUrl": "." to tsconfig.json compilerOptions
    import { WpcomRequestParams } from 'src/utils/wpcom-wrapper'; 
    export default function wpcomProxyRequest( params: WpcomRequestParams, callback: Function ): XMLHttpRequest;
}
```

File: packages/data-stores/src/user/types.ts
Comment (p-jackson) [inline packages/data-stores/src/user/types.ts]: I haven't looked the actual server response, but I'm guessing some of these properties are only optional because they don't exist when `success===false`?

In which case the data can be modeled like this:
```typescript
interface NewUserSuccessResponse {
	success: true;
	is_signup_sandbox: boolean;
	bearer_token: string | null;

	// I guess these would still have to be optional since only one is defined I think.
	username?: string;
	signup_sandbox_username?: string;
	user_id?: number;
	signup_sandbox_user_id?: string;
}

interface NewUserSuccessResponse {
	success: false;
	is_signup_sandbox: boolean;
}

export type NewUserResponse = NewUserSuccessResponse | NewUserErrorResponse;
```

That means that after checking that `success===true` you won't need to do undefined checks on all of those properties.

On the other hand, now that we have `.?` syntax maybe having all of these optional properties isn't so bothersome anymore.

**Review by p-jackson — COMMENTED** [review]
I like it! This is much clearer to follow than the way the `user` singleton does it's magic in Calypso, so a huge step forward IMO.

File: packages/data-stores/src/user/constants.ts
Comment (andrewserong) [reply packages/data-stores/src/user/constants.ts]: I couldn't decide, either!

File: packages/data-stores/src/user/index.ts
Comment (andrewserong) [reply packages/data-stores/src/user/index.ts]: Nice! Sounds great.

File: packages/data-stores/src/user/controls.ts
Comment (ramonjd) [inline packages/data-stores/src/user/controls.ts]: **TIL** about controls! Thank you! 

What would be the recommendation to "chain" HTTP requests?

For example, the mega rough plan is, once the user has handed over their email address, to:

- provision an account (using the work in this PR)
- create a site straight away

⚠️ n00b question

Is that something (an action?) we could dispatch from within a controls option object? E.g.,

```
PROVISION_USER_ACCOUNT_AND_SITE: async () => {
    ....
    const newUser = await userRequest();
    newUser?.id && dispatchSomeAction...
   ...
```

**Comment by andrewserong** [comment]
Thanks @p-jackson that gives me heaps to go on, all great suggestions 👍And yes, you correctly identified that some of the oddities are carried over from the other stores :)

File: packages/data-stores/src/user/controls.ts
Comment (andrewserong) [reply packages/data-stores/src/user/controls.ts]: I think that would probably work, but I was also wondering if it'd make sense for the site to live in its own store called something like `currentSite`? Then we'd be able to chain them together in Gutenboarding instead of in controls. I was imagining it could be something like:

In Gutenboarding:
* A user selects a design and page template(s), then clicks on "create site"
* A function then checks if there's a current user, and if so fires off `createSite`.
* If not, display the account creation form, and when it's submitted, fire off `createAccount`, and on success, fire off `createSite`. To check for success we could use a similar selector as to the hacky ["New user created!"](https://github.com/automattic/wp-calypso/blob/2b6e1888311f7b4659c806c16e35d20eb89b643b/client/landing/gutenboarding/components/header/index.tsx#L140) I've added in this PR.
* `createSite` creates the site, and on success stores the current site in the data store (so that we can deal with the duplicate sites issue further down the track), then redirect the user to the editor.

File: packages/data-stores/src/utils/wpcom-wrapper.ts
Comment (andrewserong) [reply packages/data-stores/src/utils/wpcom-wrapper.ts]: Oh, I like that idea. I'll have a play with that, too.

File: packages/data-stores/src/user/index.ts
Comment (andrewserong) [reply packages/data-stores/src/user/index.ts]: Good call on this one, I've moved it over now to `packages/data-stores/src/utils/wpcom-wrapper.ts` and it feels neater :)

File: packages/data-stores/src/user/types.ts
Comment (ramonjd) [reply packages/data-stores/src/user/types.ts]: `bearer_token` should in theory always be a string too if `success === true`

**Comment by ramonjd** [comment]
I know this is WIP so forgive me if you've already thought of it, but is it worth splitting up the PR eventually into `data-stores` vs `landing/gutenboarding` stuff?

File: packages/data-stores/src/user/controls.ts
Comment (sirreal) [reply packages/data-stores/src/user/controls.ts]: > What would be the recommendation to "chain" HTTP requests?

On the technical level, this should be supported by `@wordpress/redux-routine` which is what enables the following patterns you'll see in all of our stores that are making API requests:

https://github.com/Automattic/wp-calypso/blob/0ac61f225a4e49d310f9348c7b02c2e36a60b320/packages/data-stores/src/verticals-templates/resolvers.ts#L11-L20

I believe you _could_ chain things right there:

```ts
function* doTheThings() {
  const { userId } = yield postUserCreateRequest();
  const { siteId } = yield postSiteRequest( { ownerId: userId } );
  // …
}
```

I'm not recommending it be done this way, just trying to share some familiarity with the data layer we're using.

---

> I was also wondering if it'd make sense for the site to live in its own store called something like `currentSite`?
> Then we'd be able to chain them together in Gutenboarding instead of in controls. I was imagining it could be something like […]

We've been wondering about when we should split up stores, and what the costs and benefits are. Nothing to report yet 🙂 

The benefit of breaking up should be making it easier to share a store for use without bloat, I'm not sure what the costs will be 🤔 More cumbersome interface?

The best we have to go on now is that store divisions should be made where they make sense, I think we should do our best to limit each store to s single concern.

What you outline seems reasonable. User and site creation actions are likely going to be useful in different contexts 👍

File: packages/data-stores/src/user/controls.ts
Comment (sirreal) [reply packages/data-stores/src/user/controls.ts]: > Since Gutenberg is it's own app (I think @sirreal even wants to move it to /app) I guess it needs [its] own app config?

_Gutenboarding_ not _Gutenberg_, but you're quite right that I do want to move it to its own app to facilitate and clarify that it is isolated from the rest of Calypso.

We have made an exception for `config`, but this is at the border between Calypso and Gutenboarding and not what I'd consider to be "in" the Gutenboarding app:

https://github.com/Automattic/wp-calypso/blob/b2a01ea5deb23f03daad6042c3a6d0e2bac53a92/client/landing/gutenboarding/index.tsx#L7

<details>
<summary>Aside about separate app</summary

I think using the config makes perfect sense here. The Gutenboarding "app" would really be a React component ready to be rendered, and necessary config would be passed in from Calypso in `client/landing/gutenboarding`:

```tsx
import config from '../../config';
import { Gutenboard } from '../apps/gutenboard';

window.AppBoot = () => {
	if ( ! config.isEnabled( 'gutenboarding' ) ) {
		window.location.href = '/';
	} else {
		ReactDom.render( <Gutenboarding />, document.getElementById( 'wpcom' ) );
	}
};
```

</details>

In this particular case, it's important to keep in mind that we're not talking about a separate app, but the `@auotmattic/data-stores` package that is meant to be published and shared! It won't have its own configuration, but should receive configuration provided by consumers. We haven't had to deal with that yet 🤔 

I've just noticed your comment here about accepting parameters in this store's `register`: https://github.com/Automattic/wp-calypso/pull/38823/files#r367228911 and already reached the same conclusion I was about to suggest:

> It'd be cool if `User.register()` accepted parameters so the config could be passed in when initing the store.

I was going to suggest the same thing, it seems like the most sensible place. The drawback would be having to use configuration anywhere you want to use the store. The only way to get the store key is by invoking `register()` (or knowing it and using the string, but we'd like to discourage that 🙂). Additionally, it would be very concerning if different configurations were passed at different times, that could easily lead to surprising behavior 😱 

That could be mitigated in consuming code, but that would be up to consumers to get it right:

```ts
// user-store.ts
import values from 'config';
import { user } from '@automattic/data-stores';
export default user.register( values.foo );

// component.ts
import { useSelect } from '@wordpress/data';
import { USER_STORE } from './user-store';
function MyComponent() {
  const user = useSelect( select => select( USER_STORE ).getUser() );
  // …
}
```

File: packages/data-stores/src/user/reducer.ts
Comment (sirreal) [reply packages/data-stores/src/user/reducer.ts]: The thinking with this pattern was to get very detailed information about the actions handled by any given reducer. We're open to feedback and exploration, nothing is set in stone 🙂

File: packages/data-stores/src/user/resolvers.ts
Comment (sirreal) [inline packages/data-stores/src/user/resolvers.ts]: I believe Gutenberg patters wouldn't actually make the request here, but would yield an action representing the request. _That_ action would trigger a side-effect via the corresponding `control` and when it resolved, the generator function would resume and and typically returns an action including the requested data to be populated in the store:

https://github.com/Automattic/wp-calypso/blob/b2a01ea5deb23f03daad6042c3a6d0e2bac53a92/packages/data-stores/src/verticals/resolvers.ts#L22-L29

Where the yielded action is:

```js
export const apiFetch = ( request ) => {
	return {
		type: 'API_FETCH',
		request,
	};
};
```

And there's a `control` that takes care of the side-effect:

```js
export const controls = {
	API_FETCH( { request } ) {
		return triggerFetch( request );
	},
```

(that's all from [`@wordpress/data-controls`](https://github.com/WordPress/gutenberg/blob/master/packages/data-controls/src/index.js#L123-L126))

I'm noting this as an observation and a pattern that we've been following because it's pattern core follows. However, the benefits of all the indirection still aren't entirely clear to me.

File: packages/data-stores/src/user/types.ts
Comment (sirreal) [reply packages/data-stores/src/user/types.ts]: 💯Beautiful. These are [_discriminated unions_](https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions) if anyone is interested in details.

File: packages/data-stores/tsconfig.json
Comment (sirreal) [inline packages/data-stores/tsconfig.json]: 🤔 We could add a `@types` directory in the Calypso root to share type declarations across the monorepo. I think if we add the directory to `typeRoots` and name the file `wpcom-proxy-request.d.ts`, it would pull types from there.

**Comment by andrewserong** [comment]
> I know this is WIP so forgive me if you've already thought of it, but is it worth splitting up the PR eventually into `data-stores` vs `landing/gutenboarding` stuff?

Absolutely @ramonjd — for this PR, consider anything that's in `landing/gutenboarding` as throwaway and just there for the purposes of testing what's in `data/stores`. Once this PR is in a ready to review state, I'll strip away all the `landing/gutenboarding` stuff.

File: packages/data-stores/tsconfig.json
Comment (p-jackson) [reply packages/data-stores/tsconfig.json]: Having shared types in the monorepo would be great. If `wpcom-proxy-request` is coming to the monorepo though (pbxlJb-6v-p2#comment-51) I'd almost just wait until then and add the `*.d.ts` to it's directory.

**Comment by andrewserong** [comment]
Thanks for the feedback everyone, great suggestions all around! I'm gradually working through them all and will continue on with it on Monday.

File: packages/data-stores/src/user/actions.ts
Comment (sirreal) [reply packages/data-stores/src/user/actions.ts]: Without this annotation, the type system just knows we have a `string`:
no const

With the annotation, we get the string constant `RECEIVE_CURRENT_USER`. I've been tempted to add another `const` on the object so it becomes readonly 😅:
const

Again, I'm open to trying different things.

File: packages/data-stores/tsconfig.json
Comment (sirreal) [reply packages/data-stores/tsconfig.json]: Agreed that with the imminent ingression, this one type isn't valuable. However, I'm quite interested in discovering pain points and figuring out these patterns that empower development across the monorepo. It's likely we'll come across untyped packages again.

No need to sort it _all_ out in this PR, of course 🙂

File: packages/data-stores/src/user/constants.ts
Comment (andrewserong) [reply packages/data-stores/src/user/constants.ts]: Leaving it as `user` for the time being, we can always change it later :)

File: packages/data-stores/src/utils/wpcom-wrapper.ts
Comment (andrewserong) [reply packages/data-stores/src/utils/wpcom-wrapper.ts]: I've added this in and it _seems_ to be working well so far, but I'm sure someone with more TS knowledge might have a better idea!

File: packages/data-stores/src/user/controls.ts
Comment (andrewserong) [reply packages/data-stores/src/user/controls.ts]: I've gone with passing in params to `User.register()` for this one.

File: packages/data-stores/src/user/types.ts
Comment (andrewserong) [reply packages/data-stores/src/user/types.ts]: It turns out that an unsuccessful signup has a completely different object response and no `success` param — the `success` param _is_ included when you call the endpoint with `validate: true` where an account isn't created. So I'm not sure this discriminated union is totally necessary here, but I've left it in because it seems like it could still be useful in the future (I just haven't used it in the reducer).

File: packages/data-stores/src/user/resolvers.ts
Comment (andrewserong) [reply packages/data-stores/src/user/resolvers.ts]: Thanks, this makes sense to me — I've refactored to this kind of pattern.

**Comment by andrewserong** [comment]
Thanks again for all the reviews everyone, I've switched this Draft over to Ready for Review and have removed the hacky changes in `header.tsx` that were only there for testing. I've added a readme file with example implementation of a simple signup form that can be used for further testing if need be. I've also added in error and fetching state for creating an account.

There's certainly more that can be improved here. One issue I ran into with typings is that the `body` of `WpcomRequestParams` kind of needs to have arbitrary keys, so I've gone with a string index signature with the clunky type of `string | number | boolean` but it feels like it _might_ need to be an `any`?

I'm keen for any further feedback particularly on the types and shape of the data in the reducers. I'm also finding that it's very easy for the scope of this PR to creep. Since we don't (yet) have any consumers of this part of the codebase, I'd like to try to clean this up to a mergeable state, and for further feature-like additions, see if we can do that in follow-up PRs. Thanks!

File: packages/data-stores/src/user/actions.ts
Comment (p-jackson) [reply packages/data-stores/src/user/actions.ts]: Thanks @sirreal. The alternatives—now that I've re-examined them—aren't as good as I remember :) Either we just use string literals for action types everywhere and defined action interfaces up front (not a bad approach, has pros and cons); or we used classes as action creators. I can't find the Medium post where the idea originally came from, [but it looks like this](https://gist.github.com/p-jackson/1ca023614930f4fcf89868ea216c2335).

The action-creators-as-classes method is kinda cool, but is another thing for devs to learn, and it doesn't look like our other JS Redux code. I like the idea of keeping the onramp to TS as smooth as possible. Plus it requires a (simple) middleware since Redux only accepts POJO actions out of the box.

So I've talked myself into thinking that the `as const` is probably the best of all the worlds.

Another way to const-ify :all-the-things: is to use the `<const>` notation:

```typescript
function createActionConst( value: number ) {
	return <const>{
		type: 'MY_TYPE',
		value
	};
}
```

That seems kinda nice.

File: packages/data-stores/global.d.ts
Comment (p-jackson) [inline packages/data-stores/global.d.ts]: Having the comment right above `declare module` doesn't seem quite right. I _think_ it's safe to move the import statement out of the block so it's at the top of the file.

Comment (p-jackson) [reply packages/data-stores/global.d.ts]: I'm actually wondering if `WpcomRequestParams` belongs in this file and that the utils should import it from here 🤔 
Seems like `wpcom-proxy-request` is the package that defines that object shape.

Comment (p-jackson) [reply packages/data-stores/global.d.ts]: > I think it's safe to move the import statement out of the block

Nope, you were right, just tried it :D

File: packages/data-stores/src/user/controls.ts
Comment (p-jackson) [inline packages/data-stores/src/user/controls.ts]: Looks a bit clumsy, but should we get this from the enum since we're getting all the other type names from the enum?

```suggestion
		[ ActionType.CREATE_ACCOUNT ]: async ( action: CreateAccountAction ) => {
```

Same with the `FETCH_CURRENT_USER` key below.

**Review by p-jackson — COMMENTED** [review]
W00t thanks for writing a readme!

**Comment by p-jackson** [comment]
I think the tests will pass after a rebase (`master` was broken)

**Comment by p-jackson** [comment]
> One issue I ran into with typings is that the body of WpcomRequestParams kind of needs to have arbitrary keys, so I've gone with a string index signature with the clunky type of string | number | boolean but it feels like it might need to be an any?

I'd probably just go with `object` for the `body` property. A request body can be a nested object which wouldn't be permitted given the current approach. Otherwise going with `any` would be fine.

**Comment by sirreal** [comment]
Rebased.

File: packages/data-stores/src/user/controls.ts
Comment (andrewserong) [reply packages/data-stores/src/user/controls.ts]: Nice, that makes things a bit more consistent.

**Comment by andrewserong** [comment]
> I'd probably just go with object for the body property. A request body can be a nested object which wouldn't be permitted given the current approach.

Thanks @p-jackson that's much simpler! I've also added in using the ActionType enum for the object keys in `controls`.

@sirreal's doing [some good exploration of a refactor](https://github.com/Automattic/wp-calypso/pull/38959) of some of this PR, but I'm currently thinking it might be worth merging this one in first, and then doing refactors separately since the overall interface that'll be used by Gutenboarding remains the same (selectors, dispatch, registering the store, etc). That way we can unblock progressing with the passwordless signup form.

What does everything think? Anything else you'd like me to address here before we're good to merge? (And apologies if I've missed any of the commentary so far, I've been trying to make sure I'm factoring all the suggestions in where I can)

**Review by sirreal — COMMENTED** [review]
> @sirreal's doing [some good exploration of a refactor](https://github.com/Automattic/wp-calypso/pull/38959) of some of this PR, but I'm currently thinking it might be worth merging this one in first, and then doing refactors separately since the overall interface that'll be used by Gutenboarding remains the same (selectors, dispatch, registering the store, etc). That way we can unblock progressing with the passwordless signup form.

👍 This is fine with me, please merge when your team considers this ready, y'all are doing good work here.

My intention with #38959 is to outline the general direction I think this should move. If you agree with the direction, please make changes in that direction how and when you see fit. I don't plan to pursue the changes in the short term to I'd be happy if you adopt #38959.

I've left one note about the `isLoading…` state that's relevant. Again, you may prefer to land this and address that feedback in a follow-up. I don't want to see this PR held up any longer than is necessary.

**Comment by andrewserong** [comment]
> My intention with #38959 is to outline the general direction I think this should move. If you agree with the direction, please make changes in that direction how and when you see fit. I don't plan to pursue the changes in the short term to I'd be happy if you adopt #38959.

Thanks so much @sirreal, absolutely, I like the general direction. I'm more than happy to take over the PR and keep on making improvements as we go. I'll take a look into `isResolving` too, but I'll try to move toward making smaller incremental improvement PRs as we go :)

**Review by p-jackson — APPROVED** [review]
lgtm!

You're right I think we'll keep jamming on this in subsequent PRs. Will be great to have this building block in place!

File: packages/data-stores/src/user/types.ts
Comment (ramonjd) [inline packages/data-stores/src/user/types.ts]: Is the constant name mismatch intentional? Just saw this.

I can't see where `RECEIVE_NEW_USER_SUCCESS` is used anywhere.

**Comment by sirreal** [comment]
> I'm more than happy to take over the PR and keep on making improvements as we go.

Perfect 👍 

> I'll try to move toward making smaller incremental improvement PRs as we go :)

💯 🚀

File: packages/data-stores/src/user/types.ts
Comment (andrewserong) [reply packages/data-stores/src/user/types.ts]: I'll make a note to remove this, it was there when I thought we'd need to specifically handle a success action separately from dealing with the response data, but didn't end up needing it.

**Pull request merged**
